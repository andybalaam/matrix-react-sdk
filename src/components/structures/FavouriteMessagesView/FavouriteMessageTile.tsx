/*
Copyright 2023 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React, { FC } from "react";
import { MatrixEvent } from "matrix-js-sdk/src/models/event";

import RoomContext from "../../../contexts/RoomContext";
import SettingsStore from "../../../settings/SettingsStore";
import { RoomPermalinkCreator } from "../../../utils/permalinks/Permalinks";
import DateSeparator from "../../views/messages/DateSeparator";
import EventTile from "../../views/rooms/EventTile";
import { shouldFormContinuation } from "../MessagePanel";
import { wantsDateSeparator } from "../../../DateUtils";
import { haveRendererForEvent } from "../../../events/EventTileFactory";
import { Layout } from "../../../settings/enums/Layout";
import { FavouriteMessagesStore } from "../../../stores/FavouriteMessagesStore";

interface IProps {
    // an event result object
    result: MatrixEvent;
    // href for the highlights in this result
    resultLink: string;
    // a list of strings to be highlighted in the results
    searchHighlights?: string[];
    onHeightChanged?: () => void;
    permalinkCreator?: RoomPermalinkCreator;
    //a list containing the saved items events
    timeline: MatrixEvent[];

    // Provide this to use a different store for favourite messages
    // e.g. in tests. If not supplied, we use the global default.
    favouriteMessagesStore?: FavouriteMessagesStore;
}

const FavouriteMessageTile: FC<IProps> = (props: IProps) => {
    let context!: React.ContextType<typeof RoomContext>;

    const result = props.result;
    const eventId = result.getId();

    const ts1 = result?.getTs();
    const ret = [<DateSeparator key={ts1 + "-search"} roomId={result.getRoomId()!} ts={ts1} />];
    const layout = SettingsStore.getValue<Layout>("layout");
    const isTwelveHour = !!SettingsStore.getValue<boolean>("showTwelveHourTimestamps");
    const alwaysShowTimestamps = !!SettingsStore.getValue<boolean>("alwaysShowTimestamps");
    const threadsEnabled = !!SettingsStore.getValue<boolean>("feature_threadenabled");

    for (let j = 0; j < props?.timeline.length; j++) {
        const mxEv = props?.timeline[j];
        const highlights = props?.searchHighlights;

        if (haveRendererForEvent(mxEv, context?.showHiddenEvents)) {
            // do we need a date separator since the last event?
            const prevEv = props.timeline[j - 1];
            // is this a continuation of the previous message?
            const continuation =
                !dateSeparator(prevEv, mxEv) &&
                shouldFormContinuation(prevEv, mxEv, context?.showHiddenEvents, threadsEnabled);

            let lastInSection = true;
            const nextEv = props?.timeline[j + 1];
            if (nextEv) {
                lastInSection =
                    dateSeparator(mxEv, nextEv) ||
                    mxEv.getSender() !== nextEv.getSender() ||
                    !shouldFormContinuation(mxEv, nextEv, context?.showHiddenEvents, threadsEnabled);
            }

            ret.push(
                <EventTile
                    key={`${eventId}+${j}`}
                    mxEvent={mxEv}
                    layout={layout}
                    highlights={highlights}
                    permalinkCreator={props.permalinkCreator}
                    highlightLink={props.resultLink}
                    onHeightChanged={props.onHeightChanged}
                    isTwelveHour={isTwelveHour}
                    alwaysShowTimestamps={alwaysShowTimestamps}
                    lastInSection={lastInSection}
                    continuation={continuation}
                    isFavouritesView={true}
                    favouriteMessagesStore={props.favouriteMessagesStore}
                />,
            );
        }
    }

    return (
        <li data-scroll-tokens={eventId}>
            <ol>{ret}</ol>
        </li>
    );
};

function dateSeparator(event1: MatrixEvent | undefined, event2: MatrixEvent | undefined): boolean {
    if (!event1 || !event2) {
        return false;
    }
    const date1 = event1.getDate();
    const date2 = event2.getDate();
    if (!date1 || !date2) {
        return false;
    }
    return wantsDateSeparator(date1, date2);
}

export default FavouriteMessageTile;
