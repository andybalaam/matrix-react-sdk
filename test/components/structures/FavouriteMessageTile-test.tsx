/*
Copyright 2023 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from "react";
import { render } from "@testing-library/react";
import { EventType, MatrixEvent, MsgType } from "matrix-js-sdk/src/matrix";

import { stubClient } from "../../test-utils";
import SettingsStore from "../../../src/settings/SettingsStore";
import FavouriteMessageTile from "../../../src/components/structures/FavouriteMessagesView/FavouriteMessageTile";

describe("FavouriteMessageTile", () => {
    const userId = "@alice:server.org";
    const roomId = "!room:server.org";
    const eventBefore = new MatrixEvent({
        type: EventType.RoomMessage,
        sender: userId,
        room_id: roomId,
        content: {
            msgtype: MsgType.Text,
            body: "i am before",
        },
        event_id: "$alices_message",
        origin_server_ts: 111111111111,
    });
    const alicesEvent = new MatrixEvent({
        type: EventType.RoomMessage,
        sender: userId,
        room_id: roomId,
        content: {
            msgtype: MsgType.Text,
            body: "i am alice",
        },
        event_id: "$alices_message",
        origin_server_ts: 222222222222,
    });
    const eventAfter = new MatrixEvent({
        type: EventType.RoomMessage,
        sender: userId,
        room_id: roomId,
        content: {
            msgtype: MsgType.Text,
            body: "i am after",
        },
        event_id: "$alices_message",
        origin_server_ts: 333333333333,
    });

    beforeEach(async () => {
        jest.spyOn(SettingsStore, "getValue").mockImplementation((setting) => setting === "feature_favourite_messages");
        stubClient();
    });

    afterEach(async () => {
        jest.resetAllMocks();
    });

    it("displays the favourite content", async () => {
        const view = render(<FavouriteMessageTile result={alicesEvent} resultLink="" timeline={[alicesEvent]} />);
        view.getByText("i am alice");
        expect(view.asFragment()).toMatchSnapshot();
    });

    it("displays the favourite content within a timeline", async () => {
        const view = render(
            <FavouriteMessageTile
                result={alicesEvent}
                resultLink=""
                timeline={[eventBefore, alicesEvent, eventAfter]}
            />,
        );
        view.getByText("i am alice");
        expect(view.asFragment()).toMatchSnapshot();
    });
});
