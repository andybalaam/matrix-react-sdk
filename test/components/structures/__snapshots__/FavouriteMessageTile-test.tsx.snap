// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavouriteMessageTile displays the favourite content 1`] = `
<DocumentFragment>
  <li
    data-scroll-tokens="$alices_message"
  >
    <ol>
      <div
        aria-label="Sun, Jan 16 1977"
        class="mx_DateSeparator"
        role="separator"
        tabindex="-1"
      >
        <hr
          role="none"
        />
        <h2
          aria-hidden="true"
        >
          Sun, Jan 16 1977
        </h2>
        <hr
          role="none"
        />
      </div>
      <li
        aria-atomic="true"
        aria-live="off"
        class="mx_EventTile mx_EventTile_lastInSection"
        data-event-id="$alices_message"
        data-has-reply="false"
        data-layout="false"
        data-scroll-tokens="$alices_message"
        data-self="false"
        tabindex="-1"
      >
        <div
          class="mx_DisambiguatedProfile"
        >
          <span
            class="mx_Username_color6 mx_DisambiguatedProfile_displayName"
            dir="auto"
          >
            @alice:server.org
          </span>
        </div>
        <div
          class="mx_EventTile_line"
        >
          <a
            aria-label="00:23"
            href="#"
          />
          <div
            class="mx_MTextBody mx_EventTile_content"
          >
            <span
              class="mx_EventTile_body"
              dir="auto"
            >
              i am alice
            </span>
          </div>
          <div
            aria-label="Message Actions"
            aria-live="off"
            class="mx_MessageActionBar"
            role="toolbar"
          >
            <div
              aria-label="Favourite"
              class="mx_AccessibleButton mx_MessageActionBar_iconButton"
              data-testid="$alices_message"
              role="button"
              tabindex="0"
            >
              <div />
            </div>
            <div
              aria-expanded="false"
              aria-haspopup="true"
              aria-label="Options"
              class="mx_AccessibleButton mx_MessageActionBar_iconButton mx_MessageActionBar_optionsButton"
              role="button"
              tabindex="-1"
            >
              <div />
            </div>
          </div>
        </div>
      </li>
    </ol>
  </li>
</DocumentFragment>
`;

exports[`FavouriteMessageTile displays the favourite content within a timeline 1`] = `
<DocumentFragment>
  <li
    data-scroll-tokens="$alices_message"
  >
    <ol>
      <div
        aria-label="Sun, Jan 16 1977"
        class="mx_DateSeparator"
        role="separator"
        tabindex="-1"
      >
        <hr
          role="none"
        />
        <h2
          aria-hidden="true"
        >
          Sun, Jan 16 1977
        </h2>
        <hr
          role="none"
        />
      </div>
      <li
        aria-atomic="true"
        aria-live="off"
        class="mx_EventTile mx_EventTile_lastInSection"
        data-event-id="$alices_message"
        data-has-reply="false"
        data-layout="false"
        data-scroll-tokens="$alices_message"
        data-self="false"
        tabindex="-1"
      >
        <div
          class="mx_DisambiguatedProfile"
        >
          <span
            class="mx_Username_color6 mx_DisambiguatedProfile_displayName"
            dir="auto"
          >
            @alice:server.org
          </span>
        </div>
        <div
          class="mx_EventTile_line"
        >
          <a
            aria-label="00:11"
            href="#"
          />
          <div
            class="mx_MTextBody mx_EventTile_content"
          >
            <span
              class="mx_EventTile_body"
              dir="auto"
            >
              i am before
            </span>
          </div>
          <div
            aria-label="Message Actions"
            aria-live="off"
            class="mx_MessageActionBar"
            role="toolbar"
          >
            <div
              aria-label="Favourite"
              class="mx_AccessibleButton mx_MessageActionBar_iconButton"
              data-testid="$alices_message"
              role="button"
              tabindex="0"
            >
              <div />
            </div>
            <div
              aria-expanded="false"
              aria-haspopup="true"
              aria-label="Options"
              class="mx_AccessibleButton mx_MessageActionBar_iconButton mx_MessageActionBar_optionsButton"
              role="button"
              tabindex="-1"
            >
              <div />
            </div>
          </div>
        </div>
      </li>
      <li
        aria-atomic="true"
        aria-live="off"
        class="mx_EventTile mx_EventTile_lastInSection"
        data-event-id="$alices_message"
        data-has-reply="false"
        data-layout="false"
        data-scroll-tokens="$alices_message"
        data-self="false"
        tabindex="-1"
      >
        <div
          class="mx_DisambiguatedProfile"
        >
          <span
            class="mx_Username_color6 mx_DisambiguatedProfile_displayName"
            dir="auto"
          >
            @alice:server.org
          </span>
        </div>
        <div
          class="mx_EventTile_line"
        >
          <a
            aria-label="00:23"
            href="#"
          />
          <div
            class="mx_MTextBody mx_EventTile_content"
          >
            <span
              class="mx_EventTile_body"
              dir="auto"
            >
              i am alice
            </span>
          </div>
          <div
            aria-label="Message Actions"
            aria-live="off"
            class="mx_MessageActionBar"
            role="toolbar"
          >
            <div
              aria-label="Favourite"
              class="mx_AccessibleButton mx_MessageActionBar_iconButton"
              data-testid="$alices_message"
              role="button"
              tabindex="0"
            >
              <div />
            </div>
            <div
              aria-expanded="false"
              aria-haspopup="true"
              aria-label="Options"
              class="mx_AccessibleButton mx_MessageActionBar_iconButton mx_MessageActionBar_optionsButton"
              role="button"
              tabindex="-1"
            >
              <div />
            </div>
          </div>
        </div>
      </li>
      <li
        aria-atomic="true"
        aria-live="off"
        class="mx_EventTile mx_EventTile_lastInSection"
        data-event-id="$alices_message"
        data-has-reply="false"
        data-layout="false"
        data-scroll-tokens="$alices_message"
        data-self="false"
        tabindex="-1"
      >
        <div
          class="mx_DisambiguatedProfile"
        >
          <span
            class="mx_Username_color6 mx_DisambiguatedProfile_displayName"
            dir="auto"
          >
            @alice:server.org
          </span>
        </div>
        <div
          class="mx_EventTile_line"
        >
          <a
            aria-label="00:35"
            href="#"
          />
          <div
            class="mx_MTextBody mx_EventTile_content"
          >
            <span
              class="mx_EventTile_body"
              dir="auto"
            >
              i am after
            </span>
          </div>
          <div
            aria-label="Message Actions"
            aria-live="off"
            class="mx_MessageActionBar"
            role="toolbar"
          >
            <div
              aria-label="Favourite"
              class="mx_AccessibleButton mx_MessageActionBar_iconButton"
              data-testid="$alices_message"
              role="button"
              tabindex="0"
            >
              <div />
            </div>
            <div
              aria-expanded="false"
              aria-haspopup="true"
              aria-label="Options"
              class="mx_AccessibleButton mx_MessageActionBar_iconButton mx_MessageActionBar_optionsButton"
              role="button"
              tabindex="-1"
            >
              <div />
            </div>
          </div>
        </div>
      </li>
    </ol>
  </li>
</DocumentFragment>
`;
